### PR TITLE
Dnc segment sql optimization

### PR DESCRIPTION
**Pull request by aarohiprasad**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | N
| Automated tests included? | N
| Related user documentation PR URL |  N
| Related developer documentation PR URL | N
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
SQL optimization for DNC segment. 

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Run segments:update command, SQL query described in addressed issue is running slow
2. SQL query is using apostrophes for reason parameter

#### Steps to test this PR:
1. Load up this PR
2. Check indexes in lead_donotcontact table
3. Run migrations:migrate
4. Index for lead_id has been deleted, new index for lead_id, channel, reason has been added
5. SQL query is not using apostrophes for reason parameter

### CODE REVIEW

**Review by escopecz — CHANGES_REQUESTED** [review]
@aarohiprasad could you please add this index also to the `DoNotContact` entity? This way the index will be created just for people migrating from an older version of Mautic but if someone installs a fresh Mautic then this index won't exist for them.

File: app/bundles/LeadBundle/Entity/DoNotContact.php [hunk 1]
```diff
@@ -67,7 +67,7 @@ public static function loadMetadata(ORM\ClassMetadata $metadata): void
 
         $builder->setTable('lead_donotcontact')
             ->setCustomRepositoryClass(\Mautic\LeadBundle\Entity\DoNotContactRepository::class)
-            ->addIndex(['reason'], 'dnc_reason_search');
+            ->addIndex(['lead_id', 'channel', 'reason'], 'dnc_reason_search');
```
Comment (escopecz) [inline app/bundles/LeadBundle/Entity/DoNotContact.php]:
```suggestion
            ->addIndex(['lead_id', 'channel', 'reason'], 'leadid_reason_channel');
```
The name should be same https://github.com/mautic/mautic/pull/13413/files#diff-6a40dd21e0b34d38b0c7b9b0e5c046ca0e7de1057ec431b84f6b03dfb1a43e29R55

Comment (aarohiprasad) [reply app/bundles/LeadBundle/Entity/DoNotContact.php]: My bad, I didn't notice!
I'll fix this

**Review by escopecz — APPROVED** [review]
Looks good. Thanks! 👍

**Comment by Galokh66** [comment]
Tested and verified sql query not showing apostrophes and lead_donotcontact indexes are properly created
Screenshot 2024-03-04 at 11 04 42 PM
Screenshot 2024-03-04 at 11 05 01 PM

**Review by Galokh66 — APPROVED** [review]

**Pull request merged**